common/models/templates/src/trie-builder.ts

```diff
@@ -0,0 +1,173 @@
+  }
+
+  node.unsorted = true;
+}
+
+/**
+ * Adds an item to the internal node at a given depth.
+ * @param item
+ * @param index
+ */
+function addItemToInternalNode(node: InternalNode, item: Entry, index: number) {
+  let char = item.key[index];
+  // If an internal node is the proper site for item, it belongs under the
+  // corresponding (sentinel, internal-use) child node signifying this.
+  if(char == undefined) {
+    char = SENTINEL_CODE_UNIT;
+  }
+  if (!node.children[char]) {
+    node.children[char] = createRootNode();
+    node.values.push(char);
+  }
+  addUnsorted(node.children[char], item, index + 1);
+}
+
+function addItemToLeaf(leaf: Leaf, item: Entry) {
+  leaf.entries.push(item);
+}
+
+/**
+ * Mutates the given Leaf to turn it into an InternalNode.
+ *
+ * NOTE: the node passed in will be DESTRUCTIVELY CHANGED into a different
+ * type when passed into this function!
+ *
+ * @param depth depth of the trie at this level.
+ */
+function convertLeafToInternalNode(leaf: Leaf, depth: number): void {
+  let entries = leaf.entries;
+
+  // Alias the current node, as the desired type.
+  let internal = (<unknown> leaf) as InternalNode;
+  internal.type = 'internal';
+
+  delete (leaf as Partial<Leaf>).entries;
+  internal.values = [];
+  internal.children = {};
+
+  // Convert the old values array into the format for interior nodes.
+  for (let item of entries) {
+    let char: string;
+    if (depth < item.key.length) {
+      char = item.key[depth];
+    } else {
+      char = SENTINEL_CODE_UNIT;
+    }
+
+    if (!internal.children[char]) {
+      internal.children[char] = createRootNode();
+      internal.values.push(char);
+    }
+    addUnsorted(internal.children[char], item, depth + 1);
+  }
+
+  internal.unsorted = true;
+}
+
+/**
+ * Recursively sort the trie, in descending order of weight.
+ * @param node any node in the trie
+ */
+function sortTrie(node: Node) {
+  if (node.type === 'leaf') {
+    if (!node.unsorted) {
+      return;
+    }
+
+    node.entries.sort(function (a, b) { return b.weight - a.weight; });
+  } else {
+    // We MUST recurse and sort children before returning.
+    for (let char of node.values) {
+      sortTrie(node.children[char]);
+    }
+
+    if (!node.unsorted) {
+      return;
+    }
+
+    node.values.sort((a, b) => {
+      return node.children[b].weight - node.children[a].weight;
+    });
+  }
+
+  delete node.unsorted;
+}
+
+/**
+ * Wrapper class for the trie and its nodes.
+ */
+export class TrieBuilder extends Trie {
+  /** The total weight of the entire trie. */
+  totalWeight: number;
+
+  constructor(toKey: Wordform2Key) {
+    super(createRootNode(), 0, toKey);
+    this.totalWeight = 0;
+  }
+
+  addEntry(word: string, weight?: number) {
+    weight = (isNaN(weight ?? NaN) || !weight) ? 1 : weight;
+    this.totalWeight += weight;
+
+    addUnsorted(
+      this.root, {
+        key: this.toKey(word),
+        content: word,
+        weight: weight
+      },
+      0
+    );
+  }
```

Comment:
This is the primary new addition used to tie things together.

Whenever we add a new entry to the `Trie` root, which is done through this method, we proactively track the `weight` it contributes.

Lines 149 and 150 replace the bulk of the now-deleted `sumWeights` function.